### PR TITLE
cache initial-style jaxpr transformations

### LINKED ISSUE #3847: slow jit-compilation with odeint inside tfp.optimizer.lbfgs_minimize
I want to optimize the initial condition of a dynamical system w.r.t. a criterion based on the system's trajectory. To this end, I use `jax.experimental.ode.odeint` inside of `tfp.optimizer.lbfgs_minimize` (cf. [TFP-on-JAX L-BFGS](https://github.com/google/jax/issues/1400#issuecomment-661889919)). This might sound very particular, but I suppose this will come up more often when using JAX in the context of physics models.

However, jit-compilation is very slow, which impedes scaling-up.
Is there a good reason for slow compilation or is there a way to speed this up?

Below you see a code example and two graphs that show the compile times and the length of the code generate during jit-compilation, i.e., `len(str(jax.make_jaxpr(...)(...)))`, as a function of `odeint` integration steps. The curves have similar shape and become somewhat flat.

```python
!pip install tfp-nightly==0.12.0.dev20200723

from functools import partial
import jax
import jax.numpy as jnp
from jax.experimental.ode import odeint
from tensorflow_probability.substrates import jax as tfp


def ode_func(x, t):
  """
  Time derivative of a linear ODE that decays the state.
  """
  return -0.1 * x

def loss(x, n_integration_steps):
  """
  Minimizing this function essentially just minimizes the norm of x.
  """
  t = jnp.asarray([n*0.1 for n in range(n_integration_steps)])
  y = odeint(ode_func, x, t)
  return jnp.sum(jnp.square(y))  

@partial(jax.jit, static_argnums=(1,))
def optimize(start, n_integration_steps):
  """
  Optimizes the initial condition of the ODE system using L-BFGS.
  """
  loss_fn = partial(loss, n_integration_steps=n_integration_steps)
  optim_results = tfp.optimizer.lbfgs_minimize(
        jax.value_and_grad(loss_fn),
        initial_position=start,
        num_correction_pairs=10,
        tolerance=1e-8)
  return optim_results.converged, optim_results.position


ndims = 3
n_integration_steps = 10
start = jnp.arange(ndims, 0, -1.)
converged, results = optimize(start, n_integration_steps)
```

compile_times jaxpr_str_len

### LINKED ISSUE #9187: caching not sensitive to bound axis names not present on inputs (args or closed over)
Here's correct behavior, without `jit`:

```python
In [3]: def f():
   ...:     return lax.psum(1, 'i')
   ...:

In [6]: jax.vmap(f, axis_name='i', axis_size=2, out_axes=None)()
Out[6]: 2

In [7]: jax.vmap(f, axis_name='i', axis_size=3, out_axes=None)()
Out[7]: 3
```

But we get incorrect `jit` cache jits for this kind of computation:

```python
In [8]: f = jax.jit(f)

In [9]: jax.vmap(f, axis_name='i', axis_size=3, out_axes=None)()
Out[9]: DeviceArray(3, dtype=int32, weak_type=True)

In [10]: jax.vmap(f, axis_name='i', axis_size=2, out_axes=None)()
Out[10]: DeviceArray(3, dtype=int32, weak_type=True)
```

The issue for C++ jit caching is that the [`ThreadLocalJitState`](https://github.com/google/jax/blob/f0e4f0472d0e83f2f3ecb3f0ab31148015b7c1cd/jax/_src/config.py#L417) on which caching depends isn't updated for [changes to the trace-time named axis environment](https://github.com/google/jax/blob/f0e4f0472d0e83f2f3ecb3f0ab31148015b7c1cd/jax/core.py#L1853-L1870).

For Python jit caching the same issue happens, since [the trace context used as part of the cache key](https://github.com/google/jax/blob/f0e4f0472d0e83f2f3ecb3f0ab31148015b7c1cd/jax/linear_util.py#L266) also [is not sensitive to the axis env](https://github.com/google/jax/blob/f0e4f0472d0e83f2f3ecb3f0ab31148015b7c1cd/jax/_src/config.py#L335).

The same problem exists for the control flow primitive caches:

```python
In [21]: scanned_f = lambda _, __: (lax.psum(1, 'i'), None)

In [22]: f = lambda: lax.scan(scanned_f, 0, None, length=1)[0]

In [23]: jax.vmap(f, axis_name='i', axis_size=2, out_axes=None)()
Out[23]: DeviceArray(2, dtype=int32, weak_type=True)

In [24]: jax.vmap(f, axis_name='i', axis_size=3, out_axes=None)()
Out[24]: DeviceArray(2, dtype=int32, weak_type=True)
```

The solution there is similarly to pass in a hashable version of the trace-time axis environment as an argument to [the memoized jaxpr-forming functions](https://github.com/google/jax/blob/f0e4f0472d0e83f2f3ecb3f0ab31148015b7c1cd/jax/_src/lax/control_flow.py#L71-L118).

How did we not notice until now? Well, any kind of data dependence (on a value mapped over the named axis in question) makes the caches work. So this seems to be primarily about things like jitting/scanning simple functions involving `axis_size('i')`, `axis_index('i')`, or `random.normal(key, shape=NamedShape(3, i=4))`. Tests involving the latter is what uncovered the issue!

### PR DESCRIPTION
**Pull request by mattjj**
These changes reduce the tracing time on [the shoyer version](https://github.com/google/jax/issues/3847#issuecomment-702845696) of the code in #3847 from ~50s to ~2.5s.

~I'm a little wary to add more caching until we land #9188 to fix #9187.~ But otherwise this is straightforward!

I think we can say this fixes #3847, though there's a lot more improvement we could make in this area.

### CODE REVIEW

**Review by froystig — APPROVED** [review]

**Comment by mattjj** [comment]
This was merged, but not marked as merged...
